PMM-9590 Adding support for --credentials-source

cezmunsta opened the pull request:
To enable the use of an external credentials provider, without
the need to integrate with specific APIs, the `--credentials`
flags is being added. The end-user is then able to provide
the source file however they wish, so long as it adheres to the
following structure:

```json
{
    "username": string,
    "password": string,
    "agentpassword": string
}
```

Currently, only readable files are supported, but support for
execution may be added at a later date.

cezmunsta committed:
Adding support for --credentials

To enable the use of an external credentials provider, without
the need to integrate with specific APIs, the `--credentials`
flags is being added. The end-user is then able to provide
the source file however they wish, so long as it adheres to the
following structure:

```json
{
    "username": string,
    "password": string,
    "agentpassword": string
}
```

Currently, only readable files are supported, but support for
execution may be added at a later date.

* Added `commands.Credentials` along with its receiver functions
  `Marshal`, `Unmarshal` and `ReadFromSource`
* Refactored `commands.ReadFile` to use `content` instead of
  `certificate` to avoid confusion
* Added `commands.TestCredentials`
* Added `Credentials` to `commands/management/addMySQLCommand` to
  store the path to the credentials file
* Added `commands/management/addMySQLCommand.GetCredentials`
  receiver function to enable process the specified source. This is
  called whenever `--credentials` is set to a non-zero-length string
  and this can then be used in place of `--username`, `--password`,
  and `--agent-password`

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/pmm-admin?pullRequest=180) <br/>All committers have signed the CLA.

cezmunsta commented:
This is just present for review at the moment, supporting only the MySQL service, before proceeding with the changes for additional services

BupycHuk commented on commands/base.go:
```diff
@@ -75,6 +77,53 @@ type CommandWithContext interface {
+	if f, err := os.Lstat(src); err != nil {
+		return err
+	} else {
+		exec = f.Mode()&0111 != 0
```
Why do we check permissions?

cezmunsta replied:
This is forward planning, but currently to ensure that if an executable path is provided then the user is informed rather than thinking that it would work.

BupycHuk commented on commands/base.go:
```diff
@@ -75,6 +77,53 @@ type CommandWithContext interface {
+	}
+
+	if exec {
+		return fmt.Errorf("execution is currently unimplemented: %s is executable", src)
```
Are we going to support execution of third party files?

cezmunsta replied:
So, that would depend upon what the policy is. However, there are already tools that are executed client-side, so it is not a blanket 'no' at this point. The thought here is that instead of needing to populate the content of a file, if execution was to be supported as an option then it would allow a script to call the user's secrets manager when the path is executable.

cezmunsta replied:
@BupycHuk this has been left in place. I don't see an issue ensuring that the file is not executable even if support is never added for execution. In this scenario, it at least prevents executable files from being referenced ;)

BupycHuk replied:
It should be discussed with @rnovikovP, but I think it's not a good idea to let users execute anything to get credentials. 
If we check for permissions, then error message should be different.

cezmunsta replied:
```sh
pmm-admin add mysql @pmm-admin-flags.txt
failed to retrieve credentials from /usr/local/percona/pmm2/credentials.sh: execution is currently unimplemented: /usr/local/percona/pmm2/credentials.sh
```
@BupycHuk please explain what is wrong with the error message

BupycHuk replied:
1. why does it should mean that user tries to execute file? Please provide me an example where you use some txt file in cli command to read data and it tries to execute it?
2. `currently`? it means that we have plans to do that.

BupycHuk commented on commands/base.go:
```diff
@@ -166,12 +215,12 @@ func ReadFile(filepath string) (string, error) {
 	}
 
-	certificate, err := ioutil.ReadFile(filepath)
+	content, err := ioutil.ReadFile(filepath)
```
👍🏼

BupycHuk commented on commands/base_test.go:
```diff
@@ -34,6 +34,23 @@ func init() {
 }
 
+func TestCredentials(t *testing.T) {
+	creds := Credentials{Username: "testuser", Password: "testpass", AgentPassword: "testagentpass"}
```
I think it's better to create new json file or use string itself otherwise we are just testing `Marshal` and `Unmarshal`

cezmunsta replied:
This was to allow tests to be executed separately, but I can move the test from add_mysql_test.go to replace it as that does the full test.

BupycHuk commented on commands/base.go:
```diff
@@ -75,6 +77,53 @@ type CommandWithContext interface {
+		return err
+	}
+	return nil
+}
```
I don't think that we need these two methods.

BupycHuk commented on commands/base.go:
```diff
@@ -75,6 +77,53 @@ type CommandWithContext interface {
+	return nil
+}
+
+func (c *Credentials) ReadFromSource(src string) error {
```
it could be just a function which returns two values `(*Credentials, error)`

BupycHuk commented on commands/management/add_mysql.go:
```diff
@@ -94,6 +94,7 @@ type addMySQLCommand struct {
 	Username          string
 	Password          string
 	AgentPassword     string
+	Credentials       string
```
```suggestion
	CredentialsFile       string
```

BupycHuk commented on commands/management/add_mysql.go:
```diff
@@ -254,6 +273,7 @@ func init() {
 	AddMySQLC.Flag("username", "MySQL username").Default("root").StringVar(&AddMySQL.Username)
 	AddMySQLC.Flag("password", "MySQL password").StringVar(&AddMySQL.Password)
 	AddMySQLC.Flag("agent-password", "Custom password for /metrics endpoint").StringVar(&AddMySQL.AgentPassword)
+	AddMySQLC.Flag("credentials", "Credentials provider").ExistingFileVar(&AddMySQL.Credentials)
```
```suggestion
	AddMySQLC.Flag("credentials-file", "Path to credentials").ExistingFileVar(&AddMySQL.Credentials)
```

cezmunsta replied:
This was left as a placeholder whilst discussing accepting executables as well as files

cezmunsta replied:
I think that `credentials-src` or `credentials-source` might be more accurate. At least in my mind, using `-file` would imply the reading the file vs the possible execution and thus reading output

cezmunsta replied:
Also, `credentials-s{ource,rc}` would easily extend to support HTTP requests

cezmunsta replied:
@BupycHuk any input based upon these comments? I have left it as `credentials` for now so that they can just get replaced in one go

BupycHuk replied:
I don't think that there will be possibility to execute files, it's security risk. So let's make it `-file`

cezmunsta replied:
@BupycHuk see [my comment](https://jira.percona.com/browse/PMM-9590?focusedCommentId=295545&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-295545) on the ticket about this. I think that there are safe enough ways to do this once the plugin architecture is available, as the user could register a provider through that, allowing for signature comparison ahead of execution.

BupycHuk commented on commands/management/add_mysql_test.go:
```diff
@@ -147,6 +148,23 @@ Table statistics collection disabled (always).
+		t.Fatalf("expected 'testuser', got: %v", creds.Username)
+	}
+}
+
```
it doesn't look like this test should be part of this file

cezmunsta replied:
I will move it to replace the base test

BupycHuk commented on commands/management/add_mysql_test.go:
```diff
@@ -147,6 +148,23 @@ Table statistics collection disabled (always).
+
+	if err := creds.ReadFromSource(cr); err != nil {
+		t.Fatalf("failed to read from source: %v", err)
+	}
```
we use `github.com/stretchr/testify/assert` or `github.com/stretchr/testify/require` to assert in tests.

cezmunsta committed:
Renamed Credentials to CredentialsFile

Explicitly state the purpose of the struct member

cezmunsta committed:
Relocated tests for credentials files

cezmunsta committed:
Refactored reading of CredentialsFile

* Removed unnecessary receivers from `commands`
* Updated `commands.ReadFromSource` to return *Credentials instead
  of being a receiver
* Updated `commands.management` to accept the credentials back from
  `commands.ReadFromSource` in `addMySQLCommand.GetCredentials`
* Updated tests to use assert and require

cezmunsta committed:
Update go.sum

cezmunsta commented:
@BupycHuk I see some `staticcheck` errors in untouched code, both instances are:
`error strings should not be capitalized (ST1005)`

What would normally be done with these?

BupycHuk commented:
please use `make check-all`, it should check only changes made by you.

BupycHuk commented on commands/management/add_mysql.go:
```diff
@@ -202,6 +220,7 @@ func (cmd *addMySQLCommand) Run() (commands.Result, error) {
 			Username:       cmd.Username,
 			Password:       cmd.Password,
 			AgentPassword:  cmd.AgentPassword,
+			Credentials:    cmd.CredentialsFile,
```
I think we should override Username, Password and AgentPassword on client side, so we don't need to pass them to Server as new field

cezmunsta committed:
Minor tweaks

cezmunsta committed:
Linting etc

cezmunsta committed:
Added CredentialsFile support to external

cezmunsta committed:
Added CredentialsFile support to external serverless

cezmunsta committed:
Added CredentialsFile support to HAProxy

cezmunsta committed:
Added CredentialsFile support to MongoDB

cezmunsta committed:
Added CredentialsFile support to PostgreSQL

cezmunsta committed:
Added CredentialsFile support to ProxySQL

cezmunsta committed:
Relocated cmd.CredentialsFile condition for MySQL

cezmunsta commented on commands/base.go:
```diff
@@ -161,17 +204,17 @@ func ParseDisableCollectors(collectors string) []string {
 	}
 
-	certificate, err := ioutil.ReadFile(filepath)
+	content, err := ioutil.ReadFile(filepath.Clean(filePath))
```
@BupycHuk this was highlighted as a `gosec` issue, so I just fixed it here

cezmunsta committed:
Merge remote-tracking branch 'upstream/main' into PMM-9590

cezmunsta committed:
Use credentials-source

cezmunsta committed:
Merge remote-tracking branch 'upstream/main' into credentials

cezmunsta committed:
Updated error message for executable sources

cezmunsta commented:
Closing this PR, replacement will be generated
